src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Common/ProjectBuilder.cs

```diff
@@ -61,9 +62,15 @@ public bool Build (XamarinProject project, bool doNotCleanupOnUpdate = false, st
 			Save (project, doNotCleanupOnUpdate, saveProject);
 
 			Output = project.CreateBuildOutput (this);
-
+			
 			project.NuGetRestore (Path.Combine (XABuildPaths.TestOutputDirectory, ProjectDirectory), PackagesDirectory);
 
+			if (project.PackageReferences.Any ()) {
+				var p = Path.Combine (ProjectDirectory, "obj", "project.assets.json");
+				if (!File.Exists (p))
+					BuildInternal (Path.Combine (ProjectDirectory, project.ProjectFilePath), "Restore", parameters, environmentVariables);
```

Comment:
Ok, I think:
1. we could try the `/restore` switch and only make one MSBuild call
2. I would just run it even if `project.assets.json` exists. It doesn't take much time, and `PackageReferences` might have changed, or `TargetFrameworkVersion` could have changed.